logic/Rule.java

```diff
@@ -198,19 +196,12 @@ void validateCycles() {
      * Remove type hints in the `then` pattern that are not valid in the `when` pattern
      */
     private void pruneThenResolvedTypes() {
-        // TODO: name is inconsistent with elsewhere
         then.variables().stream().filter(variable -> variable.id().isName())
-                .forEach(thenVar ->
-                                 when.variables().stream()
-                                         .filter(whenVar -> whenVar.id().equals(thenVar.id()))
-                                         .filter(whenVar -> !(whenVar.isSatisfiable() && whenVar.resolvedTypes().isEmpty()))
-                                         .findFirst().ifPresent(whenVar -> {
-                                     if (thenVar.resolvedTypes().isEmpty() && thenVar.isSatisfiable()) {
-                                         thenVar.addResolvedTypes(whenVar.resolvedTypes());
-                                     } else thenVar.retainResolvedTypes(whenVar.resolvedTypes());
-                                     if (thenVar.resolvedTypes().isEmpty()) thenVar.setSatisfiable(false);
-                                 })
-                );
+                .forEach(thenVar -> {
+                    Variable whenVar = when.variable(thenVar.id());
+                    thenVar.retainResolvedTypes(whenVar.resolvedTypes());
```

Comment:
What happens if the when variable's resolvedTypes is empty? (In the case it could be any Thing)

Reply:
This is now impossible as an empty set of types indicates failure, and need indicates "all types".